src/ray/gcs/gcs_server/object_locator.h

```diff
@@ -0,0 +1,164 @@
+#ifndef GCS_GCS_SERVER_OBJECT_LOCATOR_H
+#define GCS_GCS_SERVER_OBJECT_LOCATOR_H
+
+#include <mutex>
+#include <unordered_map>
+#include <unordered_set>
+#include "ray/common/id.h"
+#include "ray/util/logging.h"
+
+namespace ray {
+
+namespace gcs {
+
+/// \class ObjectLocationInfo
+/// This class is used to access the object's locations.
+class ObjectLocationInfo {
+ public:
+  ObjectLocationInfo(const ObjectID &object_id);
+
+  ~ObjectLocationInfo();
+
+  /// Add location of object.
+  ///
+  /// \param node_id The object location that will be added.
+  void AddLocation(const ClientID &node_id);
+
+  /// Remove location of object.
+  ///
+  /// \param node_id The object location that will be removed.
+  /// \return The number of nodes left after remove.
+  size_t RemoveLocation(const ClientID &node_id);
+
+  /// Get object's locations.
+  ///
+  /// \return Object locations.
+  std::unordered_set<ClientID> GetLocations() const;
+
+ private:
+  ObjectID object_id_;
+
+  std::unordered_set<ClientID> locations_;
+};
+
+/// \class NodeHoldObjectInfo
+/// This class is used to access the object ids which held by node.
+class NodeHoldObjectInfo {
+ public:
+  NodeHoldObjectInfo(const ClientID &node_id);
+
+  ~NodeHoldObjectInfo();
+
+  /// Add the id of the object that the node holds.
+  ///
+  /// \param object_id The id of the object that will be added.
+  void AddObject(const ObjectID &object_id);
+
+  /// Add the id of the object that the node holds.
+  ///
+  /// \param object_ids The ids of objects that will be added.
+  void AddObjects(const std::unordered_set<ObjectID> &object_ids);
+
+  /// Get all object of the node .
+  std::unordered_set<ObjectID> GetObjects();
+
+  /// Remove the id of the object that the node no longer holds.
+  ///
+  /// \param object_id The id of the object that will be removed.
+  /// \return The number of objects after remove.
+  size_t RemoveObject(const ObjectID &object_id);
+
+ private:
+  ClientID node_id_;
+
+  std::unordered_set<ObjectID> object_ids_;
+};
+
+class ObjectLocator {
+ public:
+  ObjectLocator();
+
+  ~ObjectLocator();
+
+  /// Add location of objects.
+  ///
+  /// \param node_id The object location that will be added.
+  /// \param object_ids The ids of objects which location will be added.
+  void AddObjectsLocation(const ClientID &node_id,
+                          const std::unordered_set<ObjectID> &object_ids);
+
+  /// Add location of an object.
+  ///
+  /// \param object_id The id of object which location will be added.
+  /// \param node_id The object location that will be added.
+  void AddObjectLocation(const ObjectID &object_id, const ClientID &node_id);
+
+  /// Get object's locations.
+  ///
+  /// \param object_id The id of object to lookup.
+  /// \return Object locations.
+  std::unordered_set<ClientID> GetObjectLocations(const ObjectID &object_id);
```

Comment:
Can't use `const &` here. Because the `object location` may be updated. e.g.: both the object and location may be deleted.
If you need to avoid copying, consider returning `shared_ptr<ObjectLocationInfo>`. However, copying directly is not a big problem now.